OUTDATED: change(state): Add note subtree indexes to the database, and calculate subtrees for existing blocks

arya2 opened the pull request:
## Motivation

This is needed to support "spend before sync" in lightwalletd.

Closes #6953.

Depends-On #7334.

TODO: Add state validity test.

## Solution

- Add subtrees below initial tip height

## Review

Not yet ready for review.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

- Add state service requests for reading subtrees

arya2 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,181 @@
+                    start_height = end_height;
+                    subtree_index = tree_subtree_index;
+                }
+            }
```
This is wrong.

teor2345 replied:
There's an "implementation details" section of ticket #6953 that might be helpful here, the details are currently behind a disclosure triangle.
<img width="544" alt="Screenshot 2023-08-21 at 14 24 30" src="https://github.com/ZcashFoundation/zebra/assets/8951843/75025db9-7a9a-424f-b9de-8d1abc22c9a8">

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -74,6 +78,31 @@ pub struct DbFormatChangeThreadHandle {
+        (parse_expect_version("25.1.0"), Box::new(PruneTrees)),
+        (parse_expect_version("25.2.0"), Box::new(AddSubtrees)),
+    ]
+}
```
Turning the upgrades into a list is a good refactor, but large code movements should go in a separate PR, without any changes to functionality.

That makes them much easier to review and test.

teor2345 replied:
Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the `25.1` state upgrade.

arya2 replied:
> Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the 25.1 state upgrade.

I think it could help with avoiding merge conflicts if we do the upgrade for `25.2.1` in parallel to `25.1.1`.

teor2345 replied:
> > Actually, I think it might be better to delay this refactor for a while, because it will cause merge conflicts with bug fixes to the 25.1 state upgrade.
> 
> I think it could help with avoiding merge conflicts if we do the upgrade for `25.2.1` in parallel to `25.1.1`.

I'm not sure what you mean by "in parallel"". Do you mean "concurrently running in a separate thread", "in the same PR", or "in the same release"?

But it's not just merge conflicts that are a factor here, it's technical complexity risk, reviewer attention, testing, and developer time/scope. Reducing code changes helps with all of these.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -74,6 +78,31 @@ pub struct DbFormatChangeThreadHandle {
+
+    vec![
+        (parse_expect_version("25.1.0"), Box::new(PruneTrees)),
+        (parse_expect_version("25.2.0"), Box::new(AddSubtrees)),
```
Since the `25.2.0` version change is going to be merged in a previous PR, and merging that PR will run CI, we can't re-use that version in this PR. Instead, finish any changes to the format in this PR, and then increment the version again.

Normally we could delete all of the `25.2` cached states in Google Cloud, but I don't think that will work here, because it's a manual step that we can't make Mergify do. (And if we don't do that manual step before each CI run, the upgrade code in this PR will be skipped during the tests.)

```suggestion
        (parse_expect_version("25.3.0"), Box::new(AddSubtrees)),
```

arya2 replied:
> Since the 25.2.0 version change is going to be merged in a previous PR, and merging that PR will run CI, we can't re-use that version in this PR. Instead, finish any changes to the format in this PR, and then increment the version again.

The previous PR shouldn't merge into main without this first merging into it.

teor2345 replied:
Ah, then we'll need to delete all the `25.2` states before doing the final tests and merging both PRs.

In fact, we might need to do that deletion on a regular basis, to get decent CI results from both PRs.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,181 @@
+//! Adds Sapling and Orchard note commitment subtrees to database
```
Nit: The previous PR adds the subtrees, this PR fully populates them

```suggestion
//! Fully populate the Sapling and Orchard note commitment subtrees for existing blocks in the database.
```

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,181 @@
+        });
+
+        // Create an unbounded channel for sending sapling note commitment subtrees to be written
+        let (sapling_subtree_tx, sapling_subtree_rx) = mpsc::channel();
```
Overall, I feel like this design is more complicated than it needs to be. Is it necessary to launch separate threads to do this upgrade?

It could be easier to implement it without threads, and then see if the performance is acceptable. (Remember, the upgrade task itself is already running in a separate thread.)

#### Security / Reliability

Using an unbounded channel can potentially read all the trees into RAM, if the write thread happens to block. How much memory is that? If it's a large amount, this is a potential hang or crash issue.

(I think it should be ok after duplicates are removed, and the duplicate removal task itself is ok because its compare task drops most of those trees immediately.)

teor2345 committed:
zebra-chain changes from the subtree-boundaries branch

```sh
git checkout -b subtree-boundaries-zebra-chain main
git checkout origin/subtree-boundaries zebra-chain
git commit
```

teor2345 committed:
Temporarily populate new subtree fields with None - for revert

This temporary commit needs to be reverted in the next PR.

arya2 committed:
Applies suggestions from code review

arya2 committed:
removes from_repr_unchecked methods

arya2 committed:
simplifies loop

arya2 committed:
adds subtrees to zebra-state

arya2 committed:
uses split_at, from_repr, & updates state-db-upgrades.md

arya2 committed:
Update book/src/dev/state-db-upgrades.md

Co-authored-by: teor <teor@riseup.net>

teor2345 committed:
Merge branch 'main' into subtree-boundaries-zebra-chain

arya2 committed:
renames partial_subtree to subtree_data

arya2 committed:
tests that subtree serialization format

arya2 committed:
adds raw data format serialization round-trip test

arya2 committed:
decrements minor version and skips inserting subtrees in db

arya2 committed:
Revert "decrements minor version and skips inserting subtrees in db"

This reverts commit 0d974ac578730cb7bb39cb3203d2c5c903b8e214.

arya2 committed:
adds subtrees for existing blocks in state

arya2 committed:
rustfmt

arya2 committed:
fixes off by one bug

teor2345 commented on zebra-state/src/constants.rs:
```diff
@@ -48,7 +48,7 @@ pub(crate) const DATABASE_FORMAT_VERSION: u64 = 25;
 /// - changing the format of a column family in a compatible way, or
 /// - breaking changes with compatibility code in all supported Zebra versions.
-pub(crate) const DATABASE_FORMAT_MINOR_VERSION: u64 = 1;
+pub(crate) const DATABASE_FORMAT_MINOR_VERSION: u64 = 2;
```
Note that the patch version will also need to be reset to zero.

teor2345 replied:
I fixed this in PR #7437.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,168 @@
+) {
+    let mut subtree_count = 0;
+    let mut prev_tree: Option<_> = None;
+    for (height, tree) in upgrade_db.sapling_tree_by_height_range(..=initial_tip_height) {
```
Note: I split the lightwalletd back-tracking requirement into ticket #7407.

(Please don't fix it in this PR, it needs its own code and tests, and this PR is big enough already.)

teor2345 replied:
I changed the code below this code in PR #7437, so fixing it here would cause merge conflicts.

teor2345 replied:
Let's continue work on this in the ticket.

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -355,9 +396,30 @@ impl DiskWriteBatch {
         }
 
+        if let Some(subtree) = trees.sapling_subtree {
+            self.zs_insert(&sapling_subtree_cf, subtree.index, subtree.into_data());
```
Note: Since this PR writes to the database, it can't merge until after the 1.2.0 release.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs:
```diff
@@ -0,0 +1,168 @@
+    let mut subtree_count = 0;
+    let mut prev_tree: Option<_> = None;
+    for (height, tree) in upgrade_db.sapling_tree_by_height_range(..=initial_tip_height) {
+        if should_cancel_format_change.load(atomic::Ordering::Relaxed) {
+            break;
+        }
```
Note: the cancel handle got reverted to the channel implementation, because atomics don't correctly handle dropped states.

teor2345 replied:
I fixed this in PR #7437.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/tests/prop.rs:
```diff
@@ -361,6 +362,16 @@ fn roundtrip_sapling_tree_root() {
+    let _init_guard = zebra_test::init();
+
+    proptest!(|(mut val in any::<NoteCommitmentSubtreeData<sapling::tree::Node>>())| {
+        val.end = val.end.clamp(Height(0), MAX_ON_DISK_HEIGHT);
```
This might not do what you expect, because almost all the tests will be checking an `end` of `MAX_ON_DISK_HEIGHT`.

This issue is documented here, but for a string regex, rather than a clamp:
> In general, the ideal solution is to find a way to take a seed value and then use prop_map to transform it into the desired, irregular domain.
> This kind of works, but it has problems ... [it] also results in a pretty odd distribution of values. 

https://proptest-rs.github.io/proptest/proptest/tutorial/filtering.html
https://proptest-rs.github.io/proptest/proptest/tutorial/transforming-strategies.html

Instead, transform the input range so its distribution is roughly uniform:
```suggestion
        val.end.0 %= (MAX_ON_DISK_HEIGHT.0 + 1);
```

teor2345 replied:
This is now #7443

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/tests/prop.rs:
```diff
@@ -436,6 +447,16 @@ fn roundtrip_orchard_tree_root() {
+    let _init_guard = zebra_test::init();
+
+    proptest!(|(mut val in any::<NoteCommitmentSubtreeData<orchard::tree::Node>>())| {
+        val.end = val.end.clamp(Height(0), MAX_ON_DISK_HEIGHT);
```
```suggestion
        val.end.0 %= (MAX_ON_DISK_HEIGHT.0 + 1);
```

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -224,7 +260,9 @@ impl ZebraDb {
         NoteCommitmentTrees {
             sprout: self.sprout_tree(),
             sapling: self.sapling_tree(),
+            sapling_subtree: None,
             orchard: self.orchard_tree(),
+            orchard_subtree: None,
```
Looks like you might need `*_subtree_for_tip()` methods here.

You can get the tip trees with `zs_last_key_value()`, feel free to grab the `impl FromDisk for NoteCommitmentSubtreeIndex` from PR #7408.

teor2345 replied:
This is now ticket #7444

teor2345 commented:
I had to change more code than I expected in the RPC PR, so if you could minimise the changes in this PR, that would be helpful!

Then we can merge them both, and base any fixes on the `main` branch.

teor2345 commented:
@arya2 I needed the data from this PR to test the RPC method, so I copied that code across into PR #7437, and did the necessary fixes.

Let's keep working on the database changes in that PR?

I suggested some minor fixes to a test and a production method in this PR. Can you check if anything needs to be done?
Feel free to open new PRs or tickets for those fixes.
